File: front/src/api/pictures.ts
import {apiClient} from './client'
import type {
    AggregateRequest,
    AggregateResponse,
    BatchDryRun,
    BatchExifMode,
    BatchExifResult,
    CreatorEditMode,
    EditPictureResponse,
    ExifEditMode,
    ExifField,
    ExifOverrides,
    Job,
    OverrideExifResponse,
    PictureDetail,
    PictureListResponse,
    PictureSelection,
    PictureVariant,
    PresenceFilter,
    SetCreatorResponse,
    TrashFilter,
    TrashResponse,
} from '@/lib/types'

export interface UploadSlot {
    picture_id: string
    /** Present for a fresh file (PUT the bytes here); `null` for a deduplicated picture. */
    presigned_url: string | null
    /** True when the file's hash already matched an existing owned picture — no upload needed. */
    duplicate: boolean
    /** True when the matched existing picture is in the trash (it is NOT auto-restored). */
    was_deleted: boolean
}

/** One file requested in a batch presign: its name, (for dedup) its SHA-256 lowercase hex, and
 *  (for the storage-quota reservation, feature 22) its byte size. */
export interface BatchUploadFileInput {
    filename: string
    file_hash?: string
    size?: number
}

export interface CompleteUploadBody {
    mime_type?: string
    file_size?: number
    /** SHA-256 of the file (lowercase hex) — provisional ETag/dedupe key; the worker re-confirms it. */
    file_hash?: string
    width?: number
    height?: number
    captured_at?: string
    /** Optional source file creation time (feature 30 §10). The web upload leaves it unset (browsers
     *  expose no creation date); populated via WebDAV's `X-OC-CTime` instead. */
    original_file_created_at?: string
    initial_tags?: string[]
    /** Front-fixed import label (`Uploaded_YYYY_MM_DD_HH_MM`) — tags the new picture (feature 15). */
    upload_label?: string
    defer_pipeline?: boolean
}

export interface ListPicturesParams {
    page: number
    page_size: number
    sort?: string
    order?: string
    /** Comma-separated ltree paths (inclusive), combined per `match`. */
    include_tags?: string
    exclude_tags?: string
    /** Comma-separated ltree paths matched exactly (strict tag navigation). */
    exact?: string
    match?: 'all' | 'any'
    untagged?: boolean
    owned_only?: boolean
    shared_with_me?: boolean
    /** Trash-membership state: `exclude` (default) | `include` | `only`. Omitted when `exclude`. */
    trash?: TrashFilter
    captured_after?: string
    captured_before?: string
    /** Presence filters (feature 29 §4): `any` (omit) | `present` | `missing`. */
    gps?: PresenceFilter
    capture_date?: PresenceFilter
    /** OR convenience; 400 if combined with a non-`any` `gps`/`capture_date`. */
    missing_any?: boolean
    /** Proximity-sort references (feature 29 §6). */
    near_time?: string
    near_lat?: number
    near_lng?: number
    /** Date-fix mode (feature 30 §4): float undated pictures to the top of the current sort. */
    undated_first?: boolean
    thumbnail?: PictureVariant
}

export async function listPictures(params: ListPicturesParams): Promise<PictureListResponse> {
    // Build params object omitting undefined values so axios doesn't send empty keys.
    const query: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(params)) {
        if (v !== undefined) query[k] = v
    }
    const {data} = await apiClient.get<PictureListResponse>('/api/authenticated/pictures', {
        params: query,
    })
    return data
}

export async function getPicture(id: string): Promise<PictureDetail> {
    const {data} = await apiClient.get<PictureDetail>(`/api/authenticated/pictures/${id}`)
    return data
}

export async function getPictureUrl(
    id: string,
    variant: PictureVariant,
): Promise<{ url: string | null; variant: PictureVariant }> {
    // `url` is null when the requested variant has no object — a thumbnail variant on a picture with
    // no generated thumbnail (pending, or a non-thumbnailable format). The `original` always exists.
    const {data} = await apiClient.get<{ url: string | null; variant: PictureVariant }>(
        `/api/authenticated/pictures/${id}/url`,
        {params: {variant}},
    )
    return data
}

/**
 * Download a picture's **original** file. Resolves the presigned URL, then fetches
 * the bytes and saves them under the picture's filename. Falls back to opening the
 * URL in a new tab if the cross-origin fetch is blocked (e.g. S3 CORS), since the
 * `download` attribute is ignored for cross-origin links.
 */
export async function downloadOriginal(id: string, filename?: string | null): Promise<void> {
    const {url} = await getPictureUrl(id, 'original')
    if (!url) throw new Error('No download URL available')
    try {
        const res = await fetch(url)
        if (!res.ok) throw new Error(`HTTP ${res.status}`)
        const blob = await res.blob()
        const objectUrl = URL.createObjectURL(blob)
        const a = document.createElement('a')
        a.href = objectUrl
        a.download = filename || 'photo'
        document.body.appendChild(a)
        a.click()
        a.remove()
        URL.revokeObjectURL(objectUrl)
    } catch {
        window.open(url, '_blank', 'noopener')
    }
}

export async function editPicture(
    id: string,
    body: { set?: Partial<ExifOverrides>; clear?: ExifField[] },
): Promise<EditPictureResponse> {
    const {data} = await apiClient.post<EditPictureResponse>(
        `/api/authenticated/pictures/${id}/edit`,
        body,
    )
    return data
}

/** Result of {@link editReceivedExif}: the body plus the HTTP status (200 local / 202 propose). */
export interface ReceivedExifResult {
    data: OverrideExifResponse
    /** 200 for a local override, 202 when a proposal was accepted by the owner's backend. */
    status: number
}

/**
 * Edit a **received** picture's EXIF (`doc/features/10 §4.1`). Two modes:
 *
 * - `mode: 'local'` (default) — a recipient-local override. DB-only, no file reconcile, no job.
 *   `set` claims a sticky per-field override; `empty` claims it as empty/`null` (shadows a present
 *   owner value with emptiness); `clear` drops the claim so the owner's value flows through again.
 *   Always permitted. Returns `200`.
 * - `mode: 'propose'` — propose the edit to the owner, who auto-applies + re-announces so all
 *   recipients converge. Requires the incoming share to grant editing (`403` otherwise). The
 *   proposed fields' local overrides are cleared; `empty` folds into `clear`. Lands asynchronously —
 *   returns `202`.
 */
export async function editReceivedExif(
    id: string,
    body: { mode?: ExifEditMode; set?: Partial<ExifOverrides>; empty?: ExifField[]; clear?: ExifField[] },
): Promise<ReceivedExifResult> {
    const res = await apiClient.post<OverrideExifResponse>(`/api/authenticated/pictures/${id}/exif`, body)
    return {data: res.data, status: res.status}
}

/**
 * Set a picture's **creator** attribution credit (feature 26 §7). For an **owned** picture this sets
 * the owner-authoritative `creator` (`mode` ignored; `value` null/blank ⇒ reset to owner default);
 * for a **received** picture `mode: 'local'` (default) sets the recipient-local override (null/blank
 * clears it). A manual `value` beginning with `@`/`#` is rejected server-side (`400`) — the sigil
 * guard is enforced client-side too.
 */
export async function setCreator(
    id: string,
    body: { value: string | null; mode?: CreatorEditMode },
): Promise<SetCreatorResponse> {
    const {data} = await apiClient.post<SetCreatorResponse>(`/api/authenticated/pictures/${id}/creator`, body)
    return data
}

// ── Batch operations (feature 14 §6.11) ──────────────────────────────────────

/**
 * Type-aware aggregation over a selection (§4). Server-side GROUP BY / conditional aggregate,
 * so a select-all of thousands of pictures is never materialised. `sections` keeps it cheap —
 * the sidebar fetches `summary` eagerly and `tags`/`exif` only when those sections open.
 */
export async function aggregatePictures(body: AggregateRequest): Promise<AggregateResponse> {
    const {data} = await apiClient.post<AggregateResponse>('/api/authenticated/pictures/aggregate', body)
    return data
}

/** Body shared by the batch EXIF edit endpoint (`PATCH /pictures/exif`). */
export interface BatchExifBody {
    selection: PictureSelection
    set?: Partial<ExifOverrides>
    /** Fields to empty: owned pictures null the column, received-local pictures get a `null` override
     *  claim (10 §6.3). For propose-to-owner this folds into a clear. */
    empty?: ExifField[]
    clear?: ExifField[]
    mode?: BatchExifMode
    dry_run?: boolean
}

/** Batch EXIF edit over a selection (§5–§6). With `dry_run` returns the affected breakdown. */
export async function batchEditExif(body: BatchExifBody & { dry_run: true }): Promise<BatchDryRun>
export async function batchEditExif(body: BatchExifBody): Promise<BatchExifResult>
export async function batchEditExif(body: BatchExifBody): Promise<BatchExifResult | BatchDryRun> {
    const {data} = await apiClient.patch<BatchExifResult | BatchDryRun>('/api/authenticated/pictures/exif', body)
    return data
}

/** Batch soft-delete over a selection (§6). With `dry_run` returns `{ affected }` only. */
export async function batchTrash(selection: PictureSelection, dryRun = false): Promise<BatchDryRun> {
    const {data} = await apiClient.post<BatchDryRun>('/api/authenticated/pictures/trash', {selection, dry_run: dryRun})
    return data
}

/** Batch restore over a selection (§6). The selection must include the trashed rows. */
export async function batchRestore(selection: PictureSelection, dryRun = false): Promise<BatchDryRun> {
    const {data} = await apiClient.post<BatchDryRun>('/api/authenticated/pictures/restore', {selection, dry_run: dryRun})
    return data
}

/** Body for the batch creator edit endpoint (`PATCH /pictures/creator`, feature 26). */
export interface BatchCreatorBody {
    selection: PictureSelection
    /** New credit; `null`/blank resets owned pictures to the owner default and clears received overrides. */
    value: string | null
    dry_run?: boolean
}

/** Batch-set the creator over a selection (feature 26). Owned pictures get the authoritative
 *  `creator` (re-announced), received pictures the recipient-local `creator_override`. `dry_run`
 *  returns the owned (`edited`) vs received (`local_override`) breakdown. */
export async function batchSetCreator(body: BatchCreatorBody & { dry_run: true }): Promise<BatchDryRun>
export async function batchSetCreator(body: BatchCreatorBody): Promise<BatchDryRun>
export async function batchSetCreator(body: BatchCreatorBody): Promise<BatchDryRun> {
    const {data} = await apiClient.patch<BatchDryRun>('/api/authenticated/pictures/creator', body)
    return data
}

/** Soft-delete a picture the user holds (owned or received). */
export async function trashPicture(id: string): Promise<TrashResponse> {
    const {data} = await apiClient.post<TrashResponse>(`/api/authenticated/pictures/${id}/trash`)
    return data
}

/** Restore a soft-deleted picture (clears `deleted_at`). */
export async function restorePicture(id: string): Promise<TrashResponse> {
    const {data} = await apiClient.post<TrashResponse>(`/api/authenticated/pictures/${id}/restore`)
    return data
}

/**
 * Copy ("rescue") a received (or owned) picture into the caller's library as a new, independent
 * owned picture (feature 11 §3). Returns the new picture id. The bytes are copied server-side and
 * `gen_thumbnail` fills hashes/thumbnails asynchronously.
 */
export async function copyPicture(id: string): Promise<{ id: string }> {
    const {data} = await apiClient.post<{ id: string }>(`/api/authenticated/pictures/${id}/copy`)
    return data
}

/** One row of a picture's content-dedup group (feature 11 §5.5). */
export interface PictureCopy {
    id: string
    filename: string | null
    content_hash: string | null
    file_hash: string | null
    /** Dedup state: the visible survivor (`live`), the trash representative (`manual`), or hidden. */
    state: 'live' | 'manual' | 'boomerang' | 'content_dedupe' | 'deleted'
    updated_at: string
    owned: boolean
    owner_username: string | null
    owner_instance: string | null
    owner_deleted_at: string | null
    copy_source_owner_username: string | null
    copy_source_owner_instance: string | null
    copy_source_picture_id: string | null
}

/** List the content-dedup group of a picture (survivor + hidden siblings, feature 11 §5.5). */
export async function getPictureCopies(id: string): Promise<PictureCopy[]> {
    const {data} = await apiClient.get<{ copies: PictureCopy[] }>(`/api/authenticated/pictures/${id}/copies`)
    return data.copies
}

/** Make this picture the live survivor of its content-dedup group (hides the others). */
export async function keepCopy(id: string): Promise<void> {
    await apiClient.post(`/api/authenticated/pictures/${id}/copies/keep`)
}

export async function getJob(id: string): Promise<Job> {
    const {data} = await apiClient.get<Job>(`/api/authenticated/jobs/${id}`)
    return data
}

const PRESIGN_BATCH_SIZE = 100

/**
 * Batch-presign upload slots. Each file may carry its SHA-256 (`file_hash`) so the backend can
 * deduplicate against the user's existing owned pictures — a hit comes back with `duplicate: true`,
 * a `null` `presigned_url`, and the existing `picture_id`. `initialTags` are assigned to any such
 * deduplicated pictures server-side (new files get their tags later, on `complete`).
 *
 * The backend caps a single request at 100 files, so larger uploads are chunked.
 */
export async function beginUploadBatch(
    files: BatchUploadFileInput[],
    initialTags?: string[],
    uploadLabel?: string,
): Promise<UploadSlot[]> {
    const slots: UploadSlot[] = []
    for (let i = 0; i < files.length; i += PRESIGN_BATCH_SIZE) {
        const chunk = files.slice(i, i + PRESIGN_BATCH_SIZE)
        const {data} = await apiClient.post<UploadSlot[]>('/api/authenticated/pictures/uploads/batch', {
            files: chunk,
            initial_tags: initialTags?.length ? initialTags : undefined,
            upload_label: uploadLabel,
        })
        slots.push(...data)
    }
    return slots
}

export async function completeUpload(pictureId: string, body: CompleteUploadBody): Promise<{ id: string }> {
    const {data} = await apiClient.post<{ id: string }>(
        `/api/authenticated/pictures/uploads/${pictureId}/complete`,
        body,
    )
    return data
}

/**
 * Explicitly wake the tagging pipeline for the current user. Call once after a batch of uploads
 */
export async function wakePipeline(): Promise<void> {
    await apiClient.post('/api/authenticated/pictures/pipeline/wake')
}

File: back/src/services/pictures.rs
use crate::domain::picture::{Picture, PictureVersion};
use crate::domain::tag::TagPath;
use crate::infra::settings::keys;
use crate::repository::picture::{
    PictureSortField, PresenceFilter,
    SortOrder, TrashFilter,
};
use archypix_common::error::AppError;
use archypix_common::settings::Settings;
use chrono::{DateTime, NaiveDateTime, Utc};
use serde::{Deserialize, Serialize};
use std::str::FromStr;
use uuid::Uuid;

/// Selectable picture variant for presigning. Used both in list thumbnails and the per-picture URL endpoint.
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PictureVariant {
    Original,
    Small,
    Medium,
    Large,
}

impl PictureVariant {
    pub fn bucket(&self, settings: &Settings) -> String {
        match self {
            PictureVariant::Original => settings.get(keys::S3_BUCKET_PICTURES),
            PictureVariant::Small => settings.get(keys::S3_BUCKET_SMALL),
            PictureVariant::Medium => settings.get(keys::S3_BUCKET_MEDIUM),
            PictureVariant::Large => settings.get(keys::S3_BUCKET_LARGE),
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Original => "original",
            Self::Small => "small",
            Self::Medium => "medium",
            Self::Large => "large",
        }
    }

    /// A generated-thumbnail variant (small/medium/large). These only exist once the worker has
    /// generated thumbnails (`pictures.thumbnails_generated_at`), which it skips for
    /// non-thumbnailable formats (PDFs, some videos, …). The `original` always exists.
    pub fn is_thumbnail(&self) -> bool {
        !matches!(self, PictureVariant::Original)
    }
}

impl FromStr for PictureVariant {
    type Err = AppError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "original" => Ok(Self::Original),
            "small" => Ok(Self::Small),
            "medium" => Ok(Self::Medium),
            "large" => Ok(Self::Large),
            other => Err(AppError::BadRequest(format!("Unknown variant: {other}"))),
        }
    }
}

// Keep the old name as an alias so list_pictures still compiles.
pub type ThumbnailSize = PictureVariant;

#[derive(Debug, Clone, Deserialize)]
pub struct UploadMetadata {
    pub mime_type: Option<String>,
    pub file_size: Option<i64>,
    pub file_hash: Option<String>,
    pub width: Option<i32>,
    pub height: Option<i32>,
    pub exif_data: Option<serde_json::Value>,
    pub captured_at: Option<NaiveDateTime>,
    /// Optional source file creation time (feature 30 §10) for clients that can provide one. The web
    /// upload leaves it unset (browsers expose no creation date); persisted, never applied to `captured_at`.
    pub original_file_created_at: Option<NaiveDateTime>,
    pub initial_tags: Option<Vec<String>>,
    /// Front-provided import label (`Uploaded.YYYY_MM_DD_HH_MM`, fixed per batch). When set, the
    /// picture is tagged with it (feature 15). A single ltree label, validated server-side.
    pub upload_label: Option<String>,
    #[serde(default)]
    pub defer_pipeline: bool,
}

fn default_page() -> u32 {
    1
}
fn default_page_size() -> u32 {
    50
}

/// Great-circle distance in metres (haversine). Used to surface the per-row distance under a
/// `geo_near` sort — the same metric the DB orders by (feature 29 §6), so badge and order agree.
fn haversine_m(lat1: f64, lng1: f64, lat2: f64, lng2: f64) -> f64 {
    const EARTH_RADIUS_M: f64 = 6_371_000.0;
    let (a_lat, b_lat) = (lat1.to_radians(), lat2.to_radians());
    let d_lat = (lat2 - lat1).to_radians();
    let d_lng = (lng2 - lng1).to_radians();
    let a = (d_lat / 2.0).sin().powi(2) + a_lat.cos() * b_lat.cos() * (d_lng / 2.0).sin().powi(2);
    2.0 * EARTH_RADIUS_M * a.sqrt().asin()
}

#[derive(Debug, Clone, Deserialize)]
pub struct PictureListParams {
    #[serde(default = "default_page")]
    pub page: u32,
    #[serde(default = "default_page_size")]
    pub page_size: u32,
    #[serde(default)]
    pub sort: PictureSortField,
    #[serde(default)]
    pub order: SortOrder,
    /// Flat tag-set filter (§6.3). Comma-separated ltree paths; combined per `match`.
    pub include_tags: Option<String>,
    pub exclude_tags: Option<String>,
    /// One ltree path matched **exactly** (`tag_path = p`, no descendants) — the timeline's
    /// per-section scope (feature 35 §7). Combined with `include`/`exclude` per `match`.
    pub exact: Option<String>,
    /// `all` (AND) | `any` (OR) over `include_tags`. Default `all`.
    #[serde(rename = "match")]
    pub match_mode: Option<String>,
    /// `true` ⇒ pictures with no stored tag of any source. AND-ed with the tag arms, so the gallery
    /// can layer its cross-cutting include/exclude sets onto the root view too (feature 35 §7).
    #[serde(default)]
    pub untagged: bool,
    #[serde(default)]
    pub owned_only: bool,
    #[serde(default)]
    pub shared_with_me: bool,
    /// Trash-membership state: `exclude` (default) | `include` | `only` (trash view).
    #[serde(default)]
    pub trash: TrashFilter,
    pub captured_after: Option<DateTime<Utc>>,
    pub captured_before: Option<DateTime<Utc>>,
    /// Presence filters (feature 29 §4): `?gps=present|missing`, `?capture_date=present|missing`,
    /// `?missing_any=true` (the OR convenience, mutually exclusive with a per-field presence).
    #[serde(default)]
    pub gps: PresenceFilter,
    #[serde(default)]
    pub capture_date: PresenceFilter,
    #[serde(default)]
    pub missing_any: bool,
    /// Proximity-sort reference points (feature 29 §6): required by `sort=time_near` / `geo_near`.
    /// `near_time` is a **naive** instant (no offset), compared against the naive `captured_at`
    /// column — matches the `captured_at` string the client reads back from a picture detail.
    pub near_time: Option<NaiveDateTime>,
    pub near_lat: Option<f64>,
    pub near_lng: Option<f64>,
    /// Date-fix mode (feature 30 §4): float undated pictures to the top of the current sort so the
    /// user can fix them while the dated references stay scrollable below (`?undated_first=true`).
    #[serde(default)]
    pub undated_first: bool,
    pub thumbnail: Option<ThumbnailSize>,
}

#[derive(Debug, Serialize)]
pub struct PictureListItem {
    pub id: Uuid,
    pub filename: Option<String>,
    pub mime_type: Option<String>,
    pub width: Option<i32>,
    pub height: Option<i32>,
    pub captured_at: Option<NaiveDateTime>,
    pub ingested_at: NaiveDateTime,
    /// Both are `SortField`s, so the timeline buckets them client-side (feature 35 §6).
    pub updated_at: NaiveDateTime,
    pub file_size: Option<i64>,
    /// Source file modification time captured at ingest (feature 30 §10). Suggestion-only date source
    /// for the date-fix chip; `None` when unknown (most received / WebDAV rows).
    pub original_file_created_at: Option<NaiveDateTime>,
    /// Derived GPS presence (feature 29 §3): `gps_lat IS NOT NULL AND gps_lng IS NOT NULL`, for owned
    /// **and** received rows (received GPS lives in the promoted columns). Drives client-side
    /// highlight-in-context and the fix-tools grid-local anchor scan without a round-trip.
    pub has_gps: bool,
    /// Great-circle distance in metres from the `near_lat`/`near_lng` reference, populated **only**
    /// under a `geo_near` sort (feature 29 §6) so the client can show a "N km away" badge. `None`
    /// for other sorts and for ungeotagged rows. The list item never exposes raw coordinates, so
    /// this is the only way the client gets a per-picture distance.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub distance_m: Option<f64>,
    /// BlurHash string for progressive loading. `None` until the thumbnail worker runs.
    pub blurhash: Option<String>,
    /// EXIF orientation value (1–8). Thumbnails are stored in raw pixel orientation, so the
    /// client rotates them to display correctly.
    pub orientation: Option<i16>,
    pub thumbnail_url: Option<String>,
    /// `true` when this row is a picture owned by the local user; `false` for a received
    /// (shared) picture. Lets the client label/filter shared pictures.
    pub owned: bool,
    /// Original owner identity for received pictures (`@owner_username:owner_instance`); `None`
    /// for owned pictures.
    pub owner_username: Option<String>,
    pub owner_instance: Option<String>,
    /// Resolved creator credit for display (feature 26): `coalesce(creator_override, creator,
    /// owner_identity)`. Parsed by its leading sigil client-side (`@user:domain` / `#name` / plain).
    pub creator: String,
    /// Convergence of the file's embedded EXIF vs the DB row.
    pub exif_sync_status: crate::domain::picture::ExifSyncStatus,
    /// The recipient's own local soft-delete timestamp (trash view); `None` when not trashed.
    pub deleted_at: Option<NaiveDateTime>,
    /// Owner-deletion lifecycle for received pictures (09 §5.3): the owner's soft-delete timestamp
    /// and announced purge deadline. Drive the red "owner will delete this on X" badge.
    pub owner_deleted_at: Option<NaiveDateTime>,
    pub owner_purge_at: Option<NaiveDateTime>,
    /// `false` when this (cross-instance) picture's owner backend was unreachable while presigning
    /// its thumbnail (feature 28 §3.2), so the client can render a distinct "owner offline" tile
    /// rather than a generic placeholder. Always `true` for owned / same-backend / reachable owners,
    /// and for a cross-instance picture with no active token yet (a "no thumbnail" state, not an
    /// outage).
    pub owner_reachable: bool,
}

#[derive(Debug, Serialize)]
pub struct PictureListResult {
    pub total: i64,
    pub page: u32,
    pub page_size: u32,
    pub items: Vec<PictureListItem>,
}

#[derive(Debug, Serialize)]
pub struct PictureDetails {
    pub picture: Picture,
    pub versions: Vec<PictureVersion>,
}

/// One requested upload slot in a batch presign: a filename and, optionally, the client-computed
/// SHA-256 (lowercase hex) of the bytes about to be uploaded. The hash drives upload-time
/// deduplication against the user's existing owned pictures.
#[derive(Debug, Clone, Deserialize)]
pub struct BatchUploadFile {
    pub filename: String,
    pub file_hash: Option<String>,
    /// Client-declared byte size, enabling the presign-time quota reservation (feature 22 §5.3).
    /// When absent, only the coarse `at-or-over-quota` gate applies and the `complete_upload` hard
    /// check is the backstop.
    pub size: Option<i64>,
}

/// The outcome of one batch presign slot: either a fresh upload slot (PUT the bytes to
/// `presigned_url`, then `complete`), or a deduplication hit against an existing owned picture
/// (no upload needed — `picture_id` is the existing picture).
pub enum BatchUploadOutcome {
    New {
        picture_id: Uuid,
        presigned_url: String,
    },
    Duplicate {
        picture_id: Uuid,
        was_deleted: bool,
    },
}

/// Presign upload slots for a batch of files in one call. Returns one outcome per file in input
/// order.
///
/// Validate a front-provided import label and derive the three marker tag paths (wire form):
/// `(base, base.AlreadyExisting, base.AlreadyExisting.Deleted)`.
fn upload_marker_tags(label: &str) -> Result<(String, String, String), AppError> {
    let base = TagPath::parse(label, false)
        .map_err(AppError::BadRequest)?
        .as_ltree()
        .to_string();
    Ok((
        base.clone(),
        format!("{base}.AlreadyExisting"),
        format!("{base}.AlreadyExisting.Deleted"),
    ))
}

mod copy;
mod exif;
mod lifecycle;
mod listing;
mod presign;
mod upload;

pub use copy::*;
pub use exif::*;
pub use lifecycle::*;
pub use listing::*;
pub use presign::*;
pub use upload::*;
